### PR TITLE
Fix issue 16244 - Prevent an ICE with empty cast

### PR DESCRIPTION
**Pull request by LemonBoy**

### CODE REVIEW

**Comment by UplinkCoder** [comment]
cast() should never happen.

File: src/dtemplate.d [hunk 1]
```diff
@@ -4998,7 +4998,8 @@ private bool reliesOnTident(Type t, TemplateParameters* tparams = null, size_t i
         {
             //printf("CallExp.reliesOnTident('%s')\n", e.toChars());
             visit(cast(UnaExp)e);
-            if (!result)
+            // e.to can be null for `cast()`
+            if (!result && e.to)
```
Comment (UplinkCoder) [inline src/dtemplate.d]: e.error here ?

Comment (LemonBoy) [inline src/dtemplate.d]: Hm? A cast with no type is semantically valid albeit just a no-op

Comment (UplinkCoder) [inline src/dtemplate.d]: then maybe we can warn ?
`uint a = cast()12;` 
does indeed compile but I feel it's strange to allow that

**Comment by UplinkCoder** [comment]
@LemonBoy where does it say that an empty cast is semantically valid ?

**Comment by UplinkCoder** [comment]
@klickverbot I'd say we should fix the spec then this makes no sense what-so-ever.

**Comment by UplinkCoder** [comment]
@klickverbot then it is not a no-op. but rather a buitin cast-to-unqual.

**Comment by LemonBoy** [comment]
Rebased, the DAutoTest failure looks unrelated.
Can we somehow get this merged?

**Comment by UplinkCoder** [comment]
@andralex this is ready.

**Pull request merged**